### PR TITLE
bazel: Commit `BUILD.bazel` files, add lint to make sure they stay up-to-date

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR generates `BUILD.bazel` files for all of our Rust crates which allows developers to start building with Bazel!

It also adds a check during linting to make sure all of the BUILD files are up-to-date, moves the Bazel build out of nightly and into the main CI workflow, and updates the developer guide to include instructions for installing `bazelisk`.

Before merging this PR I will be sure to announce to `#eng-general` explaining how to generate BUILD files.

### Motivation

Progress towards https://github.com/MaterializeInc/materialize/issues/26796

### Tips for reviewer

This PR is split into separate commits, with the final commit being all generated files.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

File: doc/developer/guide.md [hunk 1]
```diff
@@ -64,6 +64,20 @@ sudo apt update
 sudo apt install docker docker-compose-plugin
 ```
 
+### Bazel
+
+Materialize can also optionally be built with [Bazel](https://bazel.build/). To
+do so first install [`bazelisk`](https://github.com/bazelbuild/bazelisk).
+
+On macOS you can do this with homebrew:
```
Comment (benesch) [inline doc/developer/guide.md:72]:
```suggestion
On macOS you can do this with Homebrew:
```

File: ci/test/pipeline.template.yml [hunk 2]
```diff
@@ -48,6 +48,22 @@ steps:
...
+        agents:
+          queue: builder-linux-x86_64
+
+      - id: build-aarch64-bazel
+        label: ":bazel: Build aarch64"
+        command: bin/ci-builder run stable bin/bazel-temp
+        depends_on: []
+        timeout_in_minutes: 60
+        agents:
+          queue: builder-linux-aarch64
```
Comment (benesch) [inline ci/test/pipeline.template.yml:65]: Does this mean we'll be paying double for builds in CI? Or do we expect caching to make a meaningful dent here?

File: ci/test/lint-main/checks/check-bazel.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+#
+# check-cargo.sh — check for Bazel issues (e.g., ensure all of our generated files are up to date).
```
Comment (benesch) [inline ci/test/lint-main/checks/check-bazel.sh:12]:
```suggestion
# check-bazel.sh — check for Bazel issues (e.g., ensure all of our generated files are up to date).
```

File: ci/test/lint-main/checks/check-bazel.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+cd "$(dirname "$0")/../../../.."
+
+. misc/shlib/shlib.bash
+
+try bin/bazel gen
+
+# Make sure we didn't generate any changes.
+try git diff --compact-summary --exit-code
+
+try_status_report
```
Comment (benesch) [inline ci/test/lint-main/checks/check-bazel.sh:25]: It'd be nice to print out here what to do if this fails (run `bin/bazel gen` locally, right?).

**Review by benesch — APPROVED** [review]

**Comment by benesch** [comment]
> Before merging this PR I will be sure to announce to `#eng-general` explaining how to generate BUILD files.

Can you add the explanation to the developer guide too? Also, does generating the BUILD files itself require Bazel?

File: ci/test/pipeline.template.yml (hunk 2 above)
Comment (def-) [reply ci/test/pipeline.template.yml:65]: Screenshot 2024-07-23 at 23 38 08
The Bazel build seems much cheaper, so this is not as bad as it seems.

**Review by def- — COMMENTED** [review]
I wish we could use bazel by default already for CI as well. Is there something blocking us? Do we want more people to try it out before doing the proper switch?
This also seems to increase the effort for developers, having to maintain both Cargo.toml and BUILD.bazel. Will Cargo.toml no longer be required once we switch to using Bazel exclusively or is the plan to stay with both?

File: doc/developer/guide.md (hunk 1 above)
Comment (ParkMyCar) [reply doc/developer/guide.md]: Updated!

File: ci/test/lint-main/checks/check-bazel.sh (hunk 3 above)
Comment (ParkMyCar) [reply ci/test/lint-main/checks/check-bazel.sh]: Updated!

File: ci/test/lint-main/checks/check-bazel.sh (hunk 4 above)
Comment (ParkMyCar) [reply ci/test/lint-main/checks/check-bazel.sh:29]: Yeah good idea, added this in with some hopefully helpful coloring to draw attention to it

**Comment by ParkMyCar** [comment]
> Can you add the explanation to the developer guide too?

Yeah as I typed that I realized I should add docs explaining how to generate files, updated the existing `bazel.md` and linked to it from the main `guide.md`!

> Also, does generating the BUILD files itself require Bazel?

The tool can be run via cargo, but right now I have `bin/bazel gen` wired up to run with Bazel. This way the tool gets cached and should be a lot faster to run.

File: ci/test/pipeline.template.yml (hunk 2 above)
Comment (ParkMyCar) [reply ci/test/pipeline.template.yml:65]: Caching should make a pretty meaningful dent, and once we get the `toolchains` repo setup and https://github.com/MaterializeInc/materialize/pull/28370 merged they should be even faster.

More than happy to keep this in nightly for now if cost is a concern!

**Comment by ParkMyCar** [comment]
> I wish we could use bazel by default already for CI as well. Is there something blocking us? Do we want more people to try it out before doing the proper switch?

We're almost there! Here are the known issues:

* https://github.com/MaterializeInc/materialize/issues/27033
* https://github.com/MaterializeInc/materialize/issues/27739
* https://github.com/MaterializeInc/materialize/issues/27735
* Validate that Bazel builds report the correct MZ version

None of them are too hard to solve, just need to find the time to do so.

> This also seems to increase the effort for developers, having to maintain both Cargo.toml and BUILD.bazel. Will Cargo.toml no longer be required once we switch to using Bazel exclusively or is the plan to stay with both?

Increasing effort on developers is a top concern! This is why we built `cargo-gazelle` so we can auto-generate ` BUILD.bazel` files and ideally the added overhead is nearly 0. For the foreseeable future our aim is to keep `Cargo.toml` as the source of truth so tools like `rust-analyzer` "just work" and adapt cargo-gazelle/Bazel as necessary.

**Pull request merged**